### DISCUSSION TITLE
Remotely cached assets

### DISCUSSION THREAD

**Discussion by nordewal** [withastro/roadmap#694] [Proposal]
# Summary

Allow for fast builds on clean build systems by using remotely cached assets. As an example, this can allow to use a cache for builds on Cloudflare and immensely reduce build time.

# Background & Motivation

I am using Cloudflare workers (like suggested in the Astro docs) to deploy a blog containing many image assets. For image optimization I rely on Astro. While building locally, the build is pretty fast as it can use the optimized images (assets) from cache.

However on Cloudflare every build takes a lot of time, as it needs to generate all assets from scratch (even if they existed in the previous build). 

# Goals

- Reduce build time on Cloudflare
- Allow usage of a remote cache and/or pre-populating the cache

# Example

I could imagine something like a `remoteAssetCache` config variable.
When generating an asset, Astro checks whether the asset already exist on that URL (`$httpAssetCache/$asset`) and if so, downloads it from there instead of generating it.

On Cloudflare, `$httpAssetCache` variable could then be set to `$host/_astro` and Astro would check whether the asset can be downloaded directly from the live system (into the cache).

## PoC

I created a short (hacky) PoC to see whether this would be useful ([code here](https://github.com/nordewal/uberlandpost.ch/commit/66048274394da84902332cf2828e86e17f6fdf25#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)). For simplicity, it just adds a pre- and post stage to the`build` command:

- pre-stage: download all assets of the last build  into the `cacheDir` folder
- build: as is, using the pre-populated `cacheDir`
- post-stage: save a list of all generated assets, so they can be fetched on the next build

On my blog it reduces the image generation time from ~8 minutes to ~10 seconds (incl. pre-populating the cache).

**Comment by maddsua** [comment]
I'd say that's more a problem on Cloudflare side than a lack of Astro feature, they just don't have any build caching for some unimaginable reason.
I'm currently in a process of moving my personal website to CF pages and it's not so fun seeing my "image-converter-10000" saying "yo bro you don't have any cache in here" on each build.

Actually you know what? You solution is genius! Just have the cache public and use a script to reload it into a new build. That a really smart move, I was going to just yolo that and let it build for however long it was gonna do that.

**Comment by nordewal** [comment]
Just read that Cloudflare is actually working on providing a build cache: https://blog.cloudflare.com/moderizing-cloudflare-pages-builds-toolbox/

Closing this issue, as this will be the best solution.
